lib/datadog/appsec/contrib/rack/request_middleware.rb

```diff
@@ -49,6 +54,50 @@ def call(env)
             AppSec::Event.record(*both_response.map { |_action, event| event }) if both_response.any?
 
             request_return
+          ensure
+            add_waf_runtime_tags(context)
+            request!
+          end
+
+          private
+
+          def first_request?
+            @first_request
+          end
+
+          def request!
+            @first_request = false
+          end
+
+          def active_trace
+            return unless defined?(Datadog::Tracing) && Datadog::Tracing.respond_to?(:active_span)
+
+            Datadog::Tracing.active_trace
+          end
+
+          def add_appsec_tags
+            return unless active_trace
+
+            active_trace.set_tag('_dd.appsec.enabled', 1)
+            active_trace.set_tag('_dd.runtime_family', 'ruby')
+            active_trace.set_tag('_dd.appsec.waf.version', Datadog::AppSec::WAF::VERSION::BASE_STRING)
+
+            if @processor.ruleset_info
+              active_trace.set_tag('_dd.appsec.event_rules.version', @processor.ruleset_info[:version])
+              if first_request?
+                active_trace.set_tag('_dd.appsec.event_rules.loaded', @processor.ruleset_info[:loaded].to_f)
+                active_trace.set_tag('_dd.appsec.event_rules.error_count', @processor.ruleset_info[:failed].to_f)
+                active_trace.set_tag('_dd.appsec.event_rules.errors', JSON.dump(@processor.ruleset_info[:errors]))
+                active_trace.set_tag('_dd.appsec.event_rules.addresses', JSON.dump(@processor.addresses))
+                active_trace.keep!
+              end
+            end
+          end
+
+          def add_waf_runtime_tags(context)
+            active_trace.set_tag('_dd.appsec.waf.timeouts', context.timeouts)
+            active_trace.set_tag('_dd.appsec.waf.duration', context.time / 1000.0)
+            active_trace.set_tag('_dd.appsec.waf.duration_ext', context.time_ext / 1000.0)
```

Comment:
Minor: Would it make sense to have helper methods that return time in the intended unit (it appears to me this is _microseconds_, but I may be wrong...?)

Reply:
I like having unit-typed values+type checking instead of using hungarian notation, I venture conversion methods in `Core::Utils::Time` could be a nice way to gradually annotate the code with unit information, and eventually have them return proper unit types, which would then be able to be type-checked.

Reply:
Hmm, without types it gets unwieldy, with potentially a lot of X_to_Y conversions (here, ns_to_us). I added a comment though, which, in addition to the `_ns` suffix, makes things a whole lot clearer.